Mixpanel/SelectorEvaluator.m

```diff
@@ -0,0 +1,761 @@
+//
+//  SelectorEvaluator.m
+//  Mixpanel
+//
+//  Created by Madhu Palani on 3/15/19.
+//  Copyright © 2019 Mixpanel. All rights reserved.
+//
+
+#import "SelectorEvaluator.h"
+
+NSString* const errDomain = @"SelectorEvaluatorError";
+
+NSString* const OPERATOR_KEY = @"operator";
+NSString* const CHILDREN_KEY = @"children";
+NSString* const PROPERTY_KEY = @"property";
+NSString* const VALUE_KEY = @"value";
+NSString* const EVENT_KEY = @"event";
+NSString* const LITERAL_KEY = @"literal";
+NSString* const WINDOW_KEY = @"window";
+NSString* const UNIT_KEY = @"unit";
+NSString* const HOUR_KEY = @"hour";
+NSString* const DAY_KEY = @"day";
+NSString* const WEEK_KEY = @"week";
+NSString* const MONTH_KEY = @"month";
+// Typecast operators
+NSString* const BOOLEAN_OPERATOR = @"boolean";
+NSString* const DATETIME_OPERATOR = @"datetime";
+NSString* const LIST_OPERATOR = @"list";
+NSString* const NUMBER_OPERATOR = @"number";
+NSString* const STRING_OPERATOR = @"string";
+// Binary operators
+NSString* const AND_OPERATOR = @"and";
+NSString* const OR_OPERATOR = @"or";
+NSString* const IN_OPERATOR = @"in";
+NSString* const NOT_IN_OPERATOR = @"not in";
+NSString* const PLUS_OPERATOR = @"+";
+NSString* const MINUS_OPERATOR = @"-";
+NSString* const MUL_OPERATOR = @"*";
+NSString* const DIV_OPERATOR = @"/";
+NSString* const MOD_OPERATOR = @"%";
+NSString* const EQUALS_OPERATOR = @"==";
+NSString* const NOT_EQUALS_OPERATOR = @"!=";
+NSString* const GREATER_THAN_OPERATOR = @">";
+NSString* const GREATER_THAN_EQUAL_OPERATOR = @">=";
+NSString* const LESS_THAN_OPERATOR = @"<";
+NSString* const LESS_THAN_EQUAL_OPERATOR = @"<=";
+// Unary operators
+NSString* const NOT_OPERATOR = @"not";
+NSString* const DEFINED_OPERATOR = @"defined";
+NSString* const NOT_DEFINED_OPERATOR = @"not defined";
+// Special words
+NSString* const NOW_LITERAL = @"now";
+
+int const LEFT = 0;
+int const RIGHT = 1;
+
+@implementation MPBoolean: NSObject
+- (instancetype)init: (BOOL) value
+{
+    self = [super init];
+    if (self) {
+        _value = value;
+    }
+    return self;
+}
+- (BOOL) isEqualToMPBoolean:(MPBoolean *)other {
+    return [self value] == [other value];
+}
+
+- (BOOL) isEqual:(id)object {
+    if (self == object) {
+        return YES;
+    }
+    
+    if (![object isKindOfClass:[MPBoolean class]]) {
+        return NO;
+    }
+    
+    return [self isEqualToMPBoolean:object];
+}
+
+@end
+
+@implementation SelectorEvaluator
+
++ (NSDate *) getCurrentDate {
+    return [NSDate date];
```

Comment:
you can use [NSDate date] directly, in ObjC, it's quite standard.

Reply:
The reason this function exists is so that I can override the default date in tests

Reply:
I have added some comments